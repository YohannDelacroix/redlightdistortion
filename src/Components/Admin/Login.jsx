import React from 'react'
import { useState } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import useAuth from '../../Hooks/useAuth';
import Header from "../Header/Header";
import Footer from "../Footer/Footer";
import axios from "../../api/axios";

const Login = () => {

    const { setAuth } =  useAuth();
    const [errMsg, setErrMsg] = useState('');
    const navigate = useNavigate();
    const location = useLocation();
    const from = location.state?.from?.pathname || "/";

    const handleLogin = async (e) => {
      //If password match
      e.preventDefault();
      let pwd = e.target.pwd.value;
      let username = e.target.username.value;
      console.log(username, pwd)
      
      try{
          const response = await axios.post("/auth", {user: username, pwd: pwd}, { 
            headers: {'Content-type': 'application/json'},
            withCredentials: true
          })

          if(response.status === 200){
              let accessToken = response.data.accessToken;
              console.log("Serv ok Access Token: " , accessToken);
              navigate(from, {replace: true});
              setAuth({username, accessToken})
          }
      }catch(err){
          if(!err?.response){
              setErrMsg('No Server Response');
          }
          else if(err.response?.status === 400){
              setErrMsg('Missing Username or Password');
          }
          else if(err.response?.status === 401){
              setErrMsg('Wrong Username or Password');
          }
          else{
              setErrMsg('Login Failed');
          }
          
      }
  }
  

  return (
    <div>
        <Header />
        
        <div className="admin-container">
          <form className="admin-login" onSubmit={handleLogin}>
              {errMsg && <div className="admin-login-error">{errMsg}</div>}
              <label htmlFor="username">Username:</label>
              <input  type="text" 
                      name="username" 
                      id="username" 
                      autoComplete="off" 
                      defaultValue="admin"
                      required />

              <label htmlFor="pwd">Password:</label>
              <input  type="password" 
                      name="pwd" 
                      id="pwd" 
                      autoComplete="new-password" 
                      required />
              <button type="submit">Log in</button>
          </form>
        </div>
        <Footer />
    </div>
  )
}

export default Login