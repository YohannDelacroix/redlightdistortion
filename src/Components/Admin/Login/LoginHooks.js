import React from "react";
import { useState } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import useAuth from '../../../Hooks/useAuth';
import axios from "../../../api/axios";
import { useEffect } from 'react';

export const useLogin = () => {
    const [usr, setUsr] = useState("admin");
    const [pwd, setPwd] = useState("");
    const [enableSubmit, setEnableSubmit] = useState(false);

    const { setAuth } =  useAuth();
    const [errMsg, setErrMsg] = useState('');

    const navigate = useNavigate();
    const location = useLocation();
    const from = location.state?.from?.pathname || "/";

    const handleChangeUsr = (e) => {
        setUsr(e.target.value);
    }

    const handleChangePwd = (e) => {
        setPwd(e.target.value);
    }

    useEffect( () => {
        setEnableSubmit(usr !== "" && pwd !== "")
    }, [usr, pwd])


    const handleLogin = async (e) => {
        console.log("handleLogin")
        let pwd = e.target.pwd.value;
        let username = e.target.username.value;

      
        if(!pwd || !username){
            console.log("Pwd or username undefined");
            return;
        }

        //If password match
        e.preventDefault();
        console.log(username, pwd)
        
        try{
            const response = await axios.post("/auth", {user: username, pwd: pwd}, { 
                headers: {'Content-type': 'application/json'},
                withCredentials: true
            })

            if(response.status === 200){
                let accessToken = response.data.accessToken;
                console.log("Serv ok Access Token: " , accessToken);
                navigate(from, {replace: true});
                setAuth({username, accessToken})
            }

        }catch(err){
            if(!err?.response){
                setErrMsg('No Server Response');
            }
            else if(err.response?.status === 400){
                setErrMsg('Missing Username or Password');
            }
            else if(err.response?.status === 401){
                setErrMsg('Wrong Username or Password');
            }
            else{
                setErrMsg('Login Failed');
            }
        }
    }

    return { errMsg, handleLogin, handleChangeUsr, handleChangePwd, enableSubmit, usr, pwd};
}