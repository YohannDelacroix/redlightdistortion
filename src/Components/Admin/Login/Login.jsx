import React from 'react'
import { useLogin } from "./LoginHooks";

const Login = () => {
    const { errMsg, handleLogin, handleChangePwd, handleChangeUsr, enableSubmit, usr } = useLogin();


    return (
        <div>
            <div className="admin-container">
            <form className="admin-login" onSubmit={handleLogin} data-testid="login-form">
                {errMsg && <div className="admin-login-error" datatest-id="error">{errMsg}</div>}
                <label htmlFor="username">Username:</label>
                <input  type="text" 
                        name="username" 
                        id="username" 
                        autoComplete="off" 
                        defaultValue={usr}
                        data-testid="usr"
                        onChange={handleChangeUsr}
                        required 
                        />

                <label htmlFor="pwd">Password:</label>
                <input  type="password" 
                        name="pwd" 
                        id="pwd" 
                        autoComplete="new-password" 
                        data-testid="pwd"
                        onChange={handleChangePwd}
                        required />
                <button type="submit" data-testid="login-btn" disabled={!enableSubmit}>Log in</button>
            </form>
            </div>
        </div>
    )
}

export default Login