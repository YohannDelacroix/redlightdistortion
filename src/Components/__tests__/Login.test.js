import { render, screen, cleanup } from '@testing-library/react'
import userEvent from '@testing-library/user-event';
import { MemoryRouter as Router} from 'react-router-dom';
import Login from "../Admin/Login/Login"

afterEach(() => {
    cleanup();
})

describe("Login component Tests", () => {
    it('should render component', () => {
        const { container } = render(<Router><Login /></Router>);
        expect(container).toMatchSnapshot();
    })

    it("should render the login form with one button", async () => {
        const component = render(<Router><Login /></Router>);
        const buttonList = await component.findAllByRole("button");
        expect(buttonList).toHaveLength(1);
    })

    it("should password input to have a password type", () => {
        render(<Router><Login /></Router>);
        const pwd = screen.getByTestId("pwd");
        expect(pwd).toHaveAttribute("type", "password");
    })

    it.skip("should display error if incorrect username or password", () => {
        render(<Router><Login /></Router>)
        const loginBtn = screen.getByTestId('login-btn');
        const usrInputNode = screen.getByTestId('usr');
        const pwdInputNode = screen.getByTestId('pwd');

        userEvent.type(usrInputNode, "sauce");
        userEvent.type(pwdInputNode, "carbonara");
        
        expect(usrInputNode).toHaveValue("adminsauce");
        expect(pwdInputNode).toHaveValue("carbonara")

        userEvent.click(loginBtn);
        expect(screen.getByTestId("error")).toBeInTheDocument();
    })

    it("should disable the button when one of user data is missing", () => {
        render(<Router><Login /></Router>)
        const loginBtn = screen.getByTestId('login-btn');
        expect(loginBtn).toHaveAttribute("disabled");
    })


    it.skip("should remove disable attribute from the button when all user data is provided", async () => {
        const container = render(<Router><Login /></Router>)
        
        const usrInputNode = container.getByTestId('usr');
        const pwdInputNode = container.getByTestId('pwd');

        userEvent.type(usrInputNode, "admin");
        userEvent.type(usrInputNode, "carbo");

        const loginBtn = container.getByTestId('login-btn');

        console.log("LOGIN BTN disabled ? : " , loginBtn.disabled)
        expect(loginBtn).not.toBeDisabled();
    })
})