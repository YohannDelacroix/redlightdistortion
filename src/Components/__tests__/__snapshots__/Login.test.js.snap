// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component Tests should render component 1`] = `
<div>
  <div>
    <div
      class="admin-container"
    >
      <form
        class="admin-login"
        data-testid="login-form"
      >
        <label
          for="username"
        >
          Username:
        </label>
        <input
          autocomplete="off"
          data-testid="usr"
          id="username"
          name="username"
          required=""
          type="text"
          value="admin"
        />
        <label
          for="pwd"
        >
          Password:
        </label>
        <input
          autocomplete="new-password"
          data-testid="pwd"
          id="pwd"
          name="pwd"
          required=""
          type="password"
        />
        <button
          data-testid="login-btn"
          disabled=""
          type="submit"
        >
          Log in
        </button>
      </form>
    </div>
  </div>
</div>
`;
